Add support for qual pushdown on the form `WHERE col=val1 OR col=val2 OR col=val3 OR ...`

jreyesr opened the pull request:
This PR adds support for qual pushdown when there is a `WHERE` condition of the form

```sql
SELECT *
...
WHERE col=val1 OR col=val2 OR col=val3 OR ...
```

where `col` is a column that is marked as a (possibly optional, but especially required) qual. `WHERE` conditions of that shape are converted into a qual with the same form as that generated by conditions of the form `WHERE col IN(val1, val2, val3, ...)`, which is already supported by all plugins due to the SDK [automatically unrolling a single qual with a list value into repeated List calls](https://github.com/turbot/steampipe-plugin-sdk/blob/b7c35885dbeb50d4284dc6ca4832e0c55cfefe7d/plugin/table_fetch.go#L136-L177).

Currently, such conditions aren't interpreted since OR is a BoolExpr, and the only supported BoolExpr is `NOT col`. This means that, for queries with a list of quals separated by ORs, that condition isn't pushed down to the plugin, thus causing a qual-less List call, which can fail if the column is a required qual.

Slack thread: https://turbot-community.slack.com/archives/C01UECB59A7/p1712277130098689

kaidaguerre commented:
@jreyesr this looks good. Have you tested what happens when you have complex combinations of AND and OR?

jreyesr commented:
> Have you tested what happens when you have complex combinations of AND and OR?

Yes, I have tested both cases where ANDs are inside an OR and ORs are inside an AND. If you have any more ideas about complex conditions that may trip this up, let me know and I'll test them.

For ORs inside ANDs, e.g. `x AND (y OR z)`, that is handled already because Postgres already presents those as separate elements on `plan.fdw_exprs`, so nothing new needs to happen. `x` on the one hand and `y OR z` on the other are converted into separate Quals and sent to the plugin: from what I can see, plugins have an implicit AND of all the quals that they receive, so that is handled already.

For ANDs inside ORs (e.g. `(x=1 AND y=1) OR (x=1 AND y=2)`, which you could also think of as encoding two points in a 2D plane, `(1, 1)` and `(1, 2)`), it depends on the exact form of the expression:

* If _all but one_ of the conditions inside each OR member are the same, which geometrically would represent fetching points in a 2D space that lie on the same horizontal/vertical line, or on a 3D space that lie on the same orthogonal line (e.g. above, where `x=1` is repeated on each member), then that is optimized by Postgres, before it even reaches the FDW, into the form `x=1 AND (y=1 OR y=2)`. AFAICS, this is documented and reliable behavior, see [the Postgres optimizer](https://github.com/postgres/postgres/blob/b1b13d2b524e64e3bf3538441366bdc8f6d3beda/src/backend/optimizer/prep/prepqual.c#L747-L8859), particularly the `find_duplicate_ors` function and its called functions. Postgres code calls it the "inverse OR distributive law". This will be handled as the case above, ORs inside ANDs, and will result in two Quals: one for `x=1` and another for `y=1 OR y=2`, which generates the same Qual and is therefore processed by plugins in the same way as if it had been `y IN(1, 2)`
* Otherwise (e.g. `(x=1 AND y=1) OR (x=2 AND y=2)`, which geometrically would be fetching completely independent points, that don't lie on the same line), if there are no shared expressions inside each OR member, that won't be processed, much like no ORs are processed now. The entire OR condition will be ignored and added to the `unhandledRestrictions` counter, and the FDW will try to list the entire table. It may be possible to make those expressions work too, and in general any OR conditions, by making repeated calls to the plugin, in a way that would emulate a UNION statement, passing a different member of the OR on each call (e.g. split that condition into two separate plugin calls, one with the Quals for `WHERE x=1 AND y=1` and another one for `WHERE x=2 AND y=2`. However, that would require many more changes and a lot of care, such as correctly handling repeated statements (i.e. the difference between `UNION` and `UNION ALL`), so I didn't implement it here

jreyesr commented:
@kaidaguerre Done!
